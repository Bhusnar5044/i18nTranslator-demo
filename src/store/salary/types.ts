import { Action } from 'redux';

export enum SalaryActionTypes {
    GET_SALARY_REQUEST = '@salary/GET_SALARY_REQUEST',
    GET_SALARY_SUCCESS = '@salary/GET_SALARY_SUCCESS',
    GET_SALARY_FAILED = '@salary/GET_SALARY_FAILED',
    EDIT_SALARY = '@salary/EDIT_SALARY',
}

export interface SankeyNode {
    name?: string;
    value?: number;
    children?: SankeyNode[];
}

export interface SankeyLink {
    source: string;
    target: string;
    value: number;
}

export type SankeyData = {
    nodes: SankeyNode[];
    links: SankeyLink[];
};
export interface SalaryState {
    isLoading: boolean;
    expenditure: SankeyNode;
    sankeyData: SankeyData;
}

export interface GetSalaryRequest extends Action {
    type: typeof SalaryActionTypes.GET_SALARY_REQUEST;
}

export interface GetSalarySuccess extends Action {
    payload: SankeyNode;
    type: typeof SalaryActionTypes.GET_SALARY_SUCCESS;
}

export interface GetSalaryFailed extends Action {
    type: typeof SalaryActionTypes.GET_SALARY_FAILED;
}

export interface EditSalary extends Action {
    payload: SankeyNode;
    type: typeof SalaryActionTypes.EDIT_SALARY;
}

export type SalaryActions = GetSalaryRequest | GetSalarySuccess | GetSalaryFailed | EditSalary;
