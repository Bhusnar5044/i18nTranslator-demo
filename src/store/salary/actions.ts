import {
    EditSalary,
    GetSalaryFailed,
    GetSalaryRequest,
    GetSalarySuccess,
    SalaryActionTypes,
    SankeyNode,
} from './types';

export const getSalaryRequest = (): GetSalaryRequest => ({
    type: SalaryActionTypes.GET_SALARY_REQUEST,
});

export const getSalarySuccess = (payload: SankeyNode): GetSalarySuccess => ({
    payload,
    type: SalaryActionTypes.GET_SALARY_SUCCESS,
});

export const getSalaryFailed = (): GetSalaryFailed => ({
    type: SalaryActionTypes.GET_SALARY_FAILED,
});

export const editSalary = (payload: SankeyNode): EditSalary => ({
    payload,
    type: SalaryActionTypes.EDIT_SALARY,
});
