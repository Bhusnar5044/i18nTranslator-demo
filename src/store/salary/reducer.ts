import { generateSankeyData } from '@utils';
import { Reducer } from 'react';
import { SalaryActionTypes, SalaryActions, SalaryState } from './types';

export const initialState: SalaryState = {
    isLoading: false,
    expenditure: {},
    sankeyData: { nodes: [], links: [] },
};

export const salaryReducer: Reducer<SalaryState, SalaryActions> = (state = initialState, action) => {
    switch (action.type) {
        case SalaryActionTypes.GET_SALARY_REQUEST:
            return { ...state, isLoading: true };
        case SalaryActionTypes.GET_SALARY_SUCCESS:
        case SalaryActionTypes.EDIT_SALARY:
        case SalaryActionTypes.DELETE_SALARY:
            return {
                ...state,
                isLoading: false,
                expenditure: action.payload,
                sankeyData: generateSankeyData(action.payload),
            };
        case SalaryActionTypes.GET_SALARY_FAILED:
            return { ...state, isLoading: false };

        default:
            return state;
    }
};
