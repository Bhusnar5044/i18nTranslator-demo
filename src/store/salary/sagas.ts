import { urls } from '@constant/urls';
import { fetch } from '@utils';
import { all, call, put, takeLatest } from 'redux-saga/effects';
import { SalaryActionTypes, getSalaryFailed, getSalarySuccess } from '.';

function* fetchSalarySagaSaga() {
    const { response, error } = yield call(fetch, {
        url: urls.salary,
        method: 'GET',
    });

    if (response?.data?.length) {
        yield put(getSalarySuccess(response.data[0]));
    }
    if (error) {
        console.log(error);
        yield put(getSalaryFailed());
    }
}

export function* watchSalary() {
    yield all([takeLatest(SalaryActionTypes.GET_SALARY_REQUEST, fetchSalarySagaSaga)]);
}
