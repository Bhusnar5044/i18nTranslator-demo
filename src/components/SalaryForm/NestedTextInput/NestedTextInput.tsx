import { Box, Button, Textfield } from '@Lib';
import { SankeyNode } from '@store/salary';
import React, { useCallback, useState } from 'react';

interface Props {
    node: SankeyNode;
    onSubmit: (node: SankeyNode) => void;
}

const Budget: React.FC<Props> = ({ node, onSubmit }) => {
    const [values, setValues] = useState<SankeyNode>(node);

    const handleValueChange = (newValue: number, item: SankeyNode) => {
        const newValues = { ...values };
        const setValue = (item: SankeyNode, value: number) => {
            item.value = value;
            if (item?.children) {
                item.children.forEach((child) => setValue(child, value));
            }
        };
        setValue(item, newValue);
        setValues(newValues);
    };

    const onSave = useCallback(() => {
        onSubmit(values);
    }, [onSubmit, values]);

    const renderBudgetItem = (item: SankeyNode) => {
        return (
            <Box key={item?.name} flexWrap="wrap" gap="1rem">
                <Textfield
                    type="text"
                    name={item?.name || ''}
                    placeholder={`Enter ${item?.name || ''}`}
                    label={item?.name || ''}
                    value={item?.value}
                    onChange={(e) => handleValueChange(+e.target.value, item)}
                    disabled={!!item?.children}
                />
                {item?.children && item.children.map(renderBudgetItem)}
            </Box>
        );
    };

    return (
        <Box flexDirection="column" gap="2rem" alignItems="start">
            {renderBudgetItem(values)}
            <Button onClick={onSave}>Save</Button>
        </Box>
    );
};

export default Budget;
