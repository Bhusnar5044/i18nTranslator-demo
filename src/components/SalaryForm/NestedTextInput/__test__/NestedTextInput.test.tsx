import { fireEvent, render, screen } from '@test-utils';
import * as salary from '@utils/mockData/salary.json';
import NestedTextInput from '../NestedTextInput';

const onSubmit = jest.fn();
const renderer = () => {
    return render(<NestedTextInput node={salary} onSubmit={onSubmit} />);
};

describe('SalaryForm', () => {
    it('should render properly', () => {
        const { container } = renderer();
        expect(container).toMatchSnapshot();
    });

    it('should should trigger onSubmit Event on click of save button', () => {
        renderer();
        const button = screen.getByRole('button', { name: 'Save' });
        expect(button).toBeInTheDocument();
        fireEvent.click(button);
        expect(onSubmit).toBeCalled();
    });

    it('should should render all the nested TextFields', () => {
        renderer();
        expect(screen.getByText(salary.name)).toBeInTheDocument();
        expect(screen.getByText(salary.children[0].name)).toBeInTheDocument();
        expect(screen.getByText(salary.children[0].children[0].name)).toBeInTheDocument();
        expect(screen.getByPlaceholderText(`Enter ${salary.name}`)).toBeInTheDocument();
        expect(screen.getByPlaceholderText(`Enter ${salary.children[0].name}`)).toBeInTheDocument();
        expect(screen.getByPlaceholderText(`Enter ${salary.children[0].children[0].name}`)).toBeInTheDocument();
    });

    it('should should trigger onSubmit Event with changed data', () => {
        renderer();
        const input = screen.getByPlaceholderText(`Enter ${salary.children[0].name}`);
        fireEvent.change(input, { target: { value: '9000' } });
        expect(input).toHaveValue('9000');
        const button = screen.getByRole('button', { name: 'Save' });
        expect(button).toBeInTheDocument();
        fireEvent.click(button);
        expect(onSubmit).toBeCalledWith({ ...salary, children: [{ ...salary.children[0], value: 9000 }] });
    });
});
