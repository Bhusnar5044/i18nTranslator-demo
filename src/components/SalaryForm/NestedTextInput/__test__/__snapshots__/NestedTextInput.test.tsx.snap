// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalaryForm should render properly 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  align-items: start;
  gap: 2rem;
}

.c1 {
  display: flex;
  flex-wrap: wrap;
  gap: 1rem;
}

.c5 {
  display: inline-block;
  outline: 0;
  cursor: pointer;
  text-align: center;
  border: 0;
  padding: 7px 16px;
  min-height: 36px;
  min-width: 36px;
  color: #ffffff;
  background: #008060;
  border-radius: 4px;
  font-weight: 500;
  font-size: 14px;
  box-shadow: rgba(0, 0, 0, 0.05) 0px 1px 0px 0px,rgba(0, 0, 0, 0.2) 0px -1px 0px 0px inset;
}

.c5 :hover {
  background: #006e52;
}

.c2 {
  position: relative;
}

.c4 {
  position: absolute;
  font-size: 1rem;
  left: 0;
  top: 50%;
  transform: translateY(-50%);
  background-color: white;
  color: gray;
  padding: 0 0.3rem;
  margin: 0 0.5rem;
  transition: 0.1s ease-out;
  transform-origin: left top;
  pointer-events: none;
}

.c3 {
  font-size: 1rem;
  outline: none;
  border: 1px solid gray;
  border-radius: 5px;
  padding: 0.5rem 0.7rem;
  color: gray;
  transition: 0.1s ease-out;
}

.c3:focus {
  border-color: #6200ee;
}

.c3:focus+label {
  color: #6200ee;
  top: 0;
  transform: translateY(-50%) scale(0.9);
}

.c3:not(:placeholder-shown)+label {
  top: 0;
  transform: translateY(-50%) scale(0.9);
}

<div>
  <div
    alignitems="start"
    class="c0"
    flexdirection="column"
    gap="2rem"
  >
    <div
      class="c1"
      flexwrap="wrap"
      gap="1rem"
    >
      <div
        class="c2"
      >
        <input
          class="c3"
          disabled=""
          name="income"
          placeholder="Enter income"
          type="text"
          value="820000"
        />
        <label
          class="c4"
        >
          income
        </label>
      </div>
      <div
        class="c1"
        flexwrap="wrap"
        gap="1rem"
      >
        <div
          class="c2"
        >
          <input
            class="c3"
            disabled=""
            name="budget"
            placeholder="Enter budget"
            type="text"
            value="104000"
          />
          <label
            class="c4"
          >
            budget
          </label>
        </div>
        <div
          class="c1"
          flexwrap="wrap"
          gap="1rem"
        >
          <div
            class="c2"
          >
            <input
              class="c3"
              disabled=""
              name="rent & bills"
              placeholder="Enter rent & bills"
              type="text"
              value="15500"
            />
            <label
              class="c4"
            >
              rent & bills
            </label>
          </div>
          <div
            class="c1"
            flexwrap="wrap"
            gap="1rem"
          >
            <div
              class="c2"
            >
              <input
                class="c3"
                name="rent"
                placeholder="Enter rent"
                type="text"
                value="11000"
              />
              <label
                class="c4"
              >
                rent
              </label>
            </div>
          </div>
          <div
            class="c1"
            flexwrap="wrap"
            gap="1rem"
          >
            <div
              class="c2"
            >
              <input
                class="c3"
                name="Laundry"
                placeholder="Enter Laundry"
                type="text"
                value="500"
              />
              <label
                class="c4"
              >
                Laundry
              </label>
            </div>
          </div>
          <div
            class="c1"
            flexwrap="wrap"
            gap="1rem"
          >
            <div
              class="c2"
            >
              <input
                class="c3"
                name="gas, water & electricity"
                placeholder="Enter gas, water & electricity"
                type="text"
                value="1000"
              />
              <label
                class="c4"
              >
                gas, water & electricity
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      class="c5"
      type="button"
    >
      Save
    </button>
  </div>
</div>
`;
